File: WhirlyGlobeSrc/WhirlyGlobeComponentTester/WhirlyGlobeComponentTester/ConfigViewController.m
/*
 *  ConfigViewController.m
 *  WhirlyGlobeComponentTester
 *
 *  Created by Steve Gifford on 7/23/12.
 *  Copyright 2011-2013 mousebird consulting
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

#import "ConfigViewController.h"

@implementation ConfigSection

+ (ConfigSection *)ConfigSectionWithName:(NSString *)sectionName rows:(NSDictionary *)rowDict singleSelect:(bool)select
{
    ConfigSection *cs = [[ConfigSection alloc] init];
    cs.sectionName = sectionName;
    cs.rows = [NSMutableDictionary dictionaryWithDictionary:rowDict];
    cs.singleSelect = select;
    
    return cs;
}

@end

@interface ConfigViewController ()

@end

@implementation ConfigViewController

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (bool)valueForSection:(NSString *)section row:(NSString *)row
{
    for (ConfigSection *cs in _values)
        if (![cs.sectionName compare:section])
        {
            return [cs.rows[row] boolValue];
        }
    
    return false;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    // Do any additional setup after loading the view from its nib.
    
    NSMutableArray *newValues = [NSMutableArray array];
    [newValues addObject:[ConfigSection ConfigSectionWithName:kMaplyTestCategoryBaseLayers
                                                         rows:
                          @{kMaplyTestBlank: @(NO),
                                    kMaplyTestGeographyClass: @(YES),
                                         kMaplyTestBlueMarble: @(NO),
                                   kMaplyTestStamenWatercolor: @(NO),
                                                kMaplyTestOSM: @(NO),
                                          kMaplyTestMapBoxSat: @(NO),
                                      kMaplyTestMapBoxTerrain: @(NO),
                                      kMaplyTestMapBoxRegular: @(NO),
                                           kMaplyTestQuadTest: @(NO),
                                     kMaplyTestQuadVectorTest: @(NO),
                                    kMaplyTestQuadTestAnimate: @(NO)}
                                                 singleSelect:true]];
    
    // We won't let the user do some things in terrain mode (overlay, basically)
    //  or 2D flatmap mode (shapes, basically)
    switch (_configOptions)
    {
        case ConfigOptionsAll:
            [newValues addObjectsFromArray:
             @[[ConfigSection ConfigSectionWithName:kMaplyTestCategoryOverlayLayers
                                               rows:
                @{kMaplyTestUSGSOrtho: @(NO),
                  kMaplyTestOWM: @(NO),
                  kMaplyTestForecastIO: @(NO),
                  kMaplyTestMapboxStreets: @(NO),
                  kMaplyMapzenVectors: @(NO)
                  }
                                       singleSelect:false],
               [ConfigSection ConfigSectionWithName:kMaplyTestCategoryObjects
                                               rows:
                @{kMaplyTestLabel2D: @(NO),
                  kMaplyTestLabel3D: @(NO),
                  kMaplyTestMarker2D: @(NO),
                  kMaplyTestMarker3D: @(NO),
                  kMaplyTestSticker: @(NO),
                  kMaplyTestShapeCylinder: @(NO),
                  kMaplyTestShapeSphere: @(NO),
                  kMaplyTestShapeGreatCircle: @(NO),
                  kMaplyTestShapeArrows: @(NO),
                  kMaplyTestModels: @(NO),
                  kMaplyTestCountry: @(NO),
                  kMaplyTestLoftedPoly: @(NO),
                  kMaplyTestMegaMarkers: @(NO),
                  kMaplyTestLatLon: @(NO),
                  kMaplyTestRoads: @(NO)}
                                       singleSelect:false],
               [ConfigSection ConfigSectionWithName:kMaplyTestCategoryAnimation
                                               rows:
                @{kMaplyTestAnimateSphere: @(NO)}
                                       singleSelect:false]]];
            break;
        case ConfigOptionsTerrain:
            break;
        case ConfigOptionsFlat:
        [newValues addObjectsFromArray:
         @[[ConfigSection ConfigSectionWithName:kMaplyTestCategoryOverlayLayers
                                           rows:
            @{kMaplyTestUSGSOrtho: @(NO),
              kMaplyTestOWM: @(NO),
              kMaplyTestForecastIO: @(NO),
              kMaplyTestMapboxStreets: @(NO),
              kMaplyMapzenVectors: @(NO)
              }
                                   singleSelect:false],
           [ConfigSection ConfigSectionWithName:kMaplyTestCategoryObjects
                                           rows:
            @{kMaplyTestLabel2D: @(NO),
              kMaplyTestLabel3D: @(NO),
              kMaplyTestMarker2D: @(NO),
              kMaplyTestMarker3D: @(NO),
              kMaplyTestSticker: @(NO),
              kMaplyTestCountry: @(NO),
              kMaplyTestMegaMarkers: @(NO),
              kMaplyTestLatLon: @(NO),
              kMaplyTestRoads: @(NO)}
                                   singleSelect:false]]];
            break;
    }

    [newValues addObject:
     [ConfigSection ConfigSectionWithName:kMaplyTestCategoryGestures
                                     rows:
      @{kMaplyTestNorthUp: @(NO),
                          kMaplyTestPinch: @(YES),
                         kMaplyTestRotate: @(YES)}
                             singleSelect:false]];
    [newValues addObject:
     [ConfigSection ConfigSectionWithName:kMaplyTestCategoryInternal
                    rows:
                    @{kMaplyTestCulling: @(NO),
                      kMaplyTestPerf: @(NO),
                      kMaplyTestWaitLoad: @(NO)}
                             singleSelect:false]];
    
    _values = newValues;
}

- (void)viewDidUnload
{
    [super viewDidUnload];
    // Release any retained subviews of the main view.
    // e.g. self.myOutlet = nil;
}

- (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation
{
	return YES;
}

#pragma marker - UITableView delegate and data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return [_values count];
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    if (section >= [_values count])
        return 0;

    ConfigSection *cs = _values[section];
    return cs.sectionName;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    if (section >= [_values count])
        return 0;
    
    ConfigSection *cs = _values[section];
    return [cs.rows count];
}

- (void)tableView:(UITableView *)tableView willDisplayCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (indexPath.section >= [_values count])
        return;
    
    ConfigSection *cs = _values[indexPath.section];
    if (indexPath.row >= [cs.rows count])
        return;
    
    NSArray *arr = [[cs.rows allKeys] sortedArrayUsingSelector:@selector(compare:)];
    NSString *key = arr[indexPath.row];
    
    bool selected = [cs.rows[key] boolValue];
    cell.backgroundColor = selected ? [UIColor colorWithRed:0.75 green:0.75 blue:1.0 alpha:1.0] : [UIColor whiteColor];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (indexPath.section >= [_values count])
        return nil;
    
    ConfigSection *cs = _values[indexPath.section];
    if (indexPath.row >= [cs.rows count])
        return nil;
    
    NSArray *arr = [[cs.rows allKeys]  sortedArrayUsingSelector:@selector(compare:)];
    UITableViewCell *cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:nil];
    NSString *key = arr[indexPath.row];
    cell.textLabel.text = key;
        
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (indexPath.section >= [_values count])
        return;
    ConfigSection *cs = _values[indexPath.section];
    if (indexPath.row >= [cs.rows count])
        return;
    NSArray *arr = [[cs.rows allKeys] sortedArrayUsingSelector:@selector(compare:)];
    NSString *key = arr[indexPath.row];
    
    bool selectState = [cs.rows[key] boolValue];
    if (cs.singleSelect)
    {
        // Turn everything else off and this one on
        if (!selectState)
        {
            for (NSString *theKey in [[cs.rows allKeys] sortedArrayUsingSelector:@selector(compare:)])
                cs.rows[theKey] = @(false);
            cs.rows[key] = @(true);
        }
        [tableView reloadSections:[NSIndexSet indexSetWithIndex:indexPath.section] withRowAnimation:UITableViewRowAnimationFade];
    } else {
        cs.rows[key] = [NSNumber numberWithBool:!selectState];
        [tableView reloadRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
    }
    
}

@end

File: WhirlyGlobeSrc/WhirlyGlobeSwiftTester/WhirlyGlobeSwiftTester/ConfigViewController.swift
//
//  ConfigViewController.swift
//  WhirlyGlobeSwiftTester
//
//  Created by jmnavarro on 14/09/15.
//  Copyright (c) 2015 Mousebird. All rights reserved.
//

import UIKit

// Section in the configuration panel
class ConfigSection {

	// Section name (as dispalyed to user)
	var sectionName: String

	// Entries (name,boolean) within the section
	var rows: [String:Bool]

	// If set, user can only select one of the options
	var singleSelect: Bool

	init(sectionName: String, rows: [String:Bool], singleSelect: Bool) {
		self.sectionName = sectionName
		self.rows = rows
		self.singleSelect = singleSelect
	}

	func selectAll(select: Bool) {
		for (k,v) in rows {
			rows[k] = select
		}
	}

	class func firstSelected(section: [String:Bool]) -> String? {
		for (k,v) in section {
			if v {
				return k
			}
		}

		return nil
	}
}


enum ConfigOptions {
	case All
	case Terrain
	case Flat
}

// Configuration view lets the user decide what to turn on and off
class ConfigViewController: UIViewController, UITableViewDataSource, UITableViewDelegate {

	// Dictionary reflecting the current values from the table
	var values = [ConfigSection]()

	// What we'll display in terms of user options
	var options = ConfigOptions.All

	// Return the configuration value for a section/row
	func valueForSection(section: String, row: String) -> Bool {
		return values
			.filter{
				$0.sectionName == section
			}.first?
			.rows[row] ?? false
	}


	func loadValues() {
		values.removeAll(keepCapacity: true)

		let baseLayersSection = ConfigSection(
			sectionName: kMaplyTestCategoryBaseLayers,
			rows: [
				kMaplyTestBlank: false,
				kMaplyTestGeographyClass: true,
				kMaplyTestBlueMarble: false,
				kMaplyTestStamenWatercolor: false,
				kMaplyTestOSM: false,
				kMaplyTestMapBoxSat: false,
				kMaplyTestQuadTest: false,
				kMaplyTestNightAndDay: false,
				kMaplyTestQuadVectorTest: false,
				kMaplyTestQuadTestAnimate: false,
				kMaplyTestElevation: false],
			singleSelect: true)

		values.append(baseLayersSection)

		// We won't let the user do some things in terrain mode (overlay, basically)
		//  or 2D flatmap mode (shapes, basically)
		switch options {
		case .All:
			let overlaySection = ConfigSection(
				sectionName: kMaplyTestCategoryOverlayLayers,
				rows: [
					kMaplyTestUSGSOrtho: false,
					kMaplyTestOWM: false,
					kMaplyTestForecastIO: false,
					kMaplyTestMapboxStreets: false,
					kMaplyMapzenVectors: false],
				singleSelect: false)

			let objectsSection = ConfigSection(
				sectionName: kMaplyTestCategoryObjects,
				rows: [
					kMaplyTestLabel2D: false,
					kMaplyTestLabel3D: false,
					kMaplyTestMarker2D: false,
					kMaplyTestMarker3D: false,
					kMaplyTestSticker: false,
					kMaplyTestShapeCylinder: false,
					kMaplyTestShapeSphere: false,
					kMaplyTestShapeGreatCircle: false,
					kMaplyTestShapeArrows: false,
					kMaplyTestModels: false,
					kMaplyTestCountry: false,
					kMaplyTestLoftedPoly: false,
					kMaplyTestQuadMarkers: false,
					kMaplyTestMegaMarkers: false,
					kMaplyTestLatLon: false,
					kMaplyTestRoads: false,
					kMaplyTestArcGIS: false,
					kMaplyTestStarsAndSun: false],
				singleSelect: false)

			let animationSection = ConfigSection(
				sectionName: kMaplyTestCategoryAnimation,
				rows: [
					kMaplyTestAnimateSphere: false],
				singleSelect: false)

			values += [overlaySection, objectsSection, animationSection]

		case .Terrain: ()

		case .Flat:
			let overlaySection = ConfigSection(
				sectionName: kMaplyTestCategoryOverlayLayers,
				rows: [
					kMaplyTestUSGSOrtho: false,
					kMaplyTestOWM: false,
					kMaplyTestForecastIO: false,
					kMaplyTestMapboxStreets: false,
					kMaplyMapzenVectors: false],
				singleSelect: false)

			let objectsSection = ConfigSection(
				sectionName: kMaplyTestCategoryObjects,
				rows: [
					kMaplyTestLabel2D: false,
					kMaplyTestLabel3D: false,
					kMaplyTestMarker2D: false,
					kMaplyTestMarker3D: false,
					kMaplyTestSticker: false,
					kMaplyTestCountry: false,
					kMaplyTestQuadMarkers: false,
					kMaplyTestMegaMarkers: false,
					kMaplyTestLatLon: false,
					kMaplyTestRoads: false,
					kMaplyTestArcGIS: false,
					kMaplyTestStarsAndSun: false],
				singleSelect: false)

			values += [overlaySection, objectsSection]
		}

		let gesturesSection = ConfigSection(
			sectionName: kMaplyTestCategoryGestures,
			rows: [
				kMaplyTestNorthUp: false,
				kMaplyTestPinch: true,
				kMaplyTestRotate: true],
			singleSelect: false)

		values.append(gesturesSection)

		let internalSection = ConfigSection(
			sectionName: kMaplyTestCategoryInternal,
			rows: [
				kMaplyTestCulling: false,
				kMaplyTestPerf: false,
				kMaplyTestWaitLoad: false],
			singleSelect: false)

		values.append(internalSection)
	}

	func numberOfSectionsInTableView(tableView: UITableView) -> Int {
		return values.count
	}

	func tableView(tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
		if section >= values.count {
			return nil
		}

		return values[section].sectionName
	}

	func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
		if section >= values.count {
			return 0
		}

		return values[section].rows.count
	}

	func tableView(tableView: UITableView,
		willDisplayCell cell: UITableViewCell, forRowAtIndexPath
		indexPath: NSIndexPath) {

		if indexPath.section >= values.count {
			return
		}

		let section = values[indexPath.section]

		if indexPath.row >= section.rows.count {
			return
		}

		let items = Array(section.rows.keys)
		let key = items[indexPath.row]
		let selected = section.rows[key]

		cell.backgroundColor = (selected ?? false)
			? UIColor(red: 0.75, green: 0.75, blue: 1.0, alpha: 1.0)
			: UIColor.whiteColor()
	}

	func tableView(tableView: UITableView,
		cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {

		let section = values[indexPath.section]
		let items = Array(section.rows.keys)
		let key = items[indexPath.row]
		let cell = UITableViewCell(style: .Default, reuseIdentifier: "cell")
		cell.textLabel?.text = key

		return cell
	}

	func tableView(tableView: UITableView,
		didSelectRowAtIndexPath indexPath: NSIndexPath) {

		if indexPath.section >= values.count {
			return
		}

		let section = values[indexPath.section]

		if indexPath.row >= section.rows.count {
			return
		}

		let items = Array(section.rows.keys)
		let key = items[indexPath.row]
		let selected = section.rows[key] ?? false

		if section.singleSelect {
			// Turn everything else off and this one on
			section.selectAll(false)
			section.rows[key] = true
		}
		else {
			section.rows[key] = !selected
		}
		tableView.reloadData()
	}

}


public let kMaplyTestCategoryBaseLayers = "Base Layers"

public let kMaplyTestBlank = "Blank"
public let kMaplyTestGeographyClass = "Geography Class - Local"
public let kMaplyTestBlueMarble = "NASA Blue Marble - Local"
public let kMaplyTestStamenWatercolor = "Stamen Watercolor - Remote"
public let kMaplyTestOSM = "OpenStreetMap (Mapquest) - Remote"
public let kMaplyTestMapBoxSat = "MapBox Satellite - Remote"
public let kMaplyTestMapBoxTerrain = "MapBox Terrain - Remote"
public let kMaplyTestMapBoxRegular = "MapBox Regular - Remote"
public let kMaplyTestNightAndDay = "Night/Day Images - Remote"
public let kMaplyTestQuadTest = "Quad Test Layer"
public let kMaplyTestQuadTestAnimate = "Quad Test Layer - Animated"
public let kMaplyTestQuadVectorTest = "Quad Vector Test Layer"
public let kMaplyTestElevation = "Cesium Elevation Test Layer"


// Overlay image layers
public let kMaplyTestCategoryOverlayLayers = "Overlay layers"

public let kMaplyTestUSGSOrtho = "USGS Ortho (WMS) - Remote"
public let kMaplyTestOWM = "OpenWeatherMap - Remote"
public let kMaplyTestForecastIO = "Forecast.IO Snapshot - Remote"
public let kMaplyTestMapboxStreets = "MapBox Streets Vectors - Remote"
public let kMaplyMapzenVectors = "Mapzen Vectors - Remote"

// Objects we can display
public let kMaplyTestCategoryObjects = "Maply Objects"

public let kMaplyTestLabel2D = "Labels - 2D"
public let kMaplyTestLabel3D = "Labels - 3D"
public let kMaplyTestMarker2D = "Markers - 2D"
public let kMaplyTestMarker3D = "Markers - 3D"
public let kMaplyTestSticker = "Stickers"
public let kMaplyTestShapeCylinder = "Shapes: Cylinders"
public let kMaplyTestShapeSphere = "Shapes: Spheres"
public let kMaplyTestShapeGreatCircle = "Shapes: Great Circles"
public let kMaplyTestShapeArrows = "Shapes: Arrows"
public let kMaplyTestModels = "Models"
public let kMaplyTestCountry = "Countries"
public let kMaplyTestLoftedPoly = "Lofted Polygons"
public let kMaplyTestQuadMarkers = "Quad Paging Markers"
public let kMaplyTestMegaMarkers = "Mega Markers"
public let kMaplyTestLatLon = "Lon/Lat lines"
public let kMaplyTestRoads = "SF Roads"
public let kMaplyTestArcGIS = "ArcGIS Vectors"
public let kMaplyTestStarsAndSun = "Stars and Sun"

// Animation
public let kMaplyTestCategoryAnimation = "Animation"

public let kMaplyTestAnimateSphere = "Animated Sphere"

// Gestures
public let kMaplyTestCategoryGestures = "Gestures"

public let kMaplyTestNorthUp = "Keep North Up"
public let kMaplyTestPinch = "Pinch Gesture"
public let kMaplyTestRotate = "Rotate Gesture"

// Low level
public let kMaplyTestCategoryInternal = "Internals"
public let kMaplyTestCulling = "Culling Optimization"
public let kMaplyTestPerf = "Performance Output"
public let kMaplyTestWaitLoad = "Image waitLoad"
